File: src/main/贪心算法/t738/MonotoneIncreasingDigits.java
package 贪心算法.t738;

import java.util.Arrays;

/**
 * @auther: sotowang
 * @date: 2020/4/14 16:26
 */
public class MonotoneIncreasingDigits {
    public int monotoneIncreasingDigits(int N) {
        if(N<10){
            return N;
        }
        char[] nums = String.valueOf(N).toCharArray();
        int len = nums.length;
        int index = -1;
        for(int i=len-1;i>0;i--){
            if(nums[i]<nums[i-1]){
                index = i;
                nums[i-1]--;

            }
        }
        if(index!=-1){
            Arrays.fill(nums,index,len,'9');
        }
        return Integer.parseInt(String.valueOf(nums));
    }
}

File: src/main/贪心算法/t738/monotoneIncreasingDigits.go
package t738

import "strconv"

func monotoneIncreasingDigits(N int) int {
	if N < 10 {
		return N
	}
	nums := []byte(strconv.Itoa(N))
	len, index := len(nums), -1
	for i := len - 1; i > 0; i-- {
		if nums[i] < nums[i-1] {
			index = i
			nums[i-1]--
		}
	}
	if index != -1 {
		for i := index; i < len; i++ {
			nums[i] = '9'
		}
	}
	res, _ := strconv.Atoi(string(nums))
	return res
}
